meetings/2021-04-28.md

```diff
@@ -0,0 +1,362 @@
+defined as:
+
+```Turtle
+:authorizes owl:inverseOf [
+    owl:propertyChainAxiom(:accessControl acl:accessTo)
+]
+```
+
+This relation, which is already implicit in ACL, as it can be defined purely in terms of those concepts, allows one to get similar behavior to the one put forward as a distinguishing feature of ACP.
+
+We then looked at the ShEx examples and showed that with a very simple inferencing step -- adding one
+triple to the graph after following an `:accessControl` `Link` header relation -- one could even apply that WAC ShEx to that graph and get clients written to the old ACL ontology to continue working.
+
+But the question then was what do the `allOf`, `anyOf`, `noneOf` and `apply*` relations bring to the picture. How do they work?
+
+We were about to start on that when time ran out.
+We had just pasted this code into the chat.
+
+```typescript
+/**
+ * # The apply policy algorithm
+ *
+ * A policy is applied if:
+ * - At least one matcher it uses matches the given context; and
+ * - All the conditions it defines are satisfied.
+ *
+ * Policy conditions behave like intersection, union and exclusion operators. In
+ * order to be satisfied:
+ * - allOf requires all of its matchers to match the given context;
+ * - anyOf requires one of its matchers to match the given context;
+ * - noneOf requires none of its matchers to match the given context.
+ *
+ * Note: Given that the noneOf condition excludes matches, a policy without a
+ * satisfied allOf or anyOf condition never applies.
+ *
+ * @param policy The policy to evaluate
+ * @param context The context in which the policy is evaluated
+ * @returns True if the policy applies, false otherwise
+ */
+export function applyPolicy(policy: IPolicy, context: IContext): boolean {
+  const matchContext = (matcher: IMatcher) => {
+    return matcher.match(context);
+  };
+
+  // A policy without positive match doesn't apply
+  if (policy.allOf.length + policy.anyOf.length === 0) {
+    return false;
+  }
+
+  return (
+    (policy.allOf.length === 0 || policy.allOf.every(matchContext)) &&
+    (policy.anyOf.length === 0 || policy.anyOf.some(matchContext)) &&
+    (policy.noneOf.length === 0 || !policy.noneOf.some(matchContext))
+  );
+}
+```
+
+**Henry**: before we look at that code, can we look at some simple RDF examples of a policy using that vocabulary?
+  
+**Matthieu** 
+
+Yes, I can write it out here.
+
+```Turtle
+prefix ex: <https://example.com/>
+
+ex:PolicyX
+    acp:allow acl:Read ; # would only having acp:allow acl:Read imply anything other than Read? Why would acp:deny be necessary?
+    acp:deny acl:Write ;
+    acp:allOf ex:AgentMatcher1, ex:IssuerMatcher1, ex:TimeMatcher1, ex:CredentialIssueTimeMatcher1 .
+
+ex:AgentMatcher1
+    acp:agent ex:Bob, ex:Alice .
+
+ex:IssuerMatcher1
+    acp:issuer ex:IdP1 .
+
+ex:TimeMatcher1
+    acp:time [
+    some owl time range e.g. sunday 10am to 2pm
+    ]
+```
+
```

Comment:
```suggestion

**Matthieu :**
Response to Sarven's comment on the necessity of `acp:deny`:
> Allowing read doesn't imply anything else than read.
> However being able to forbid access following sets of rules could be seen as useful because:
> -  You can design Policies that deny certain access following set of conditions that are as intricate as required.
> -  You can design standard Policies that deny access and can be widely applied.
>
> To be clear, an access mode is only allowed if at least one policy that applies allows it and none of the policies that apply deny it.
```